nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/OAuth2Object.java

```diff
@@ -0,0 +1,165 @@
+import org.nuxeo.ecm.platform.oauth2.providers.NuxeoOAuth2ServiceProvider;
+import org.nuxeo.ecm.platform.oauth2.providers.OAuth2ServiceProvider;
+import org.nuxeo.ecm.platform.oauth2.providers.OAuth2ServiceProviderRegistry;
+import org.nuxeo.ecm.platform.oauth2.tokens.NuxeoOAuth2Token;
+import org.nuxeo.ecm.webengine.model.WebObject;
+import org.nuxeo.ecm.webengine.model.impl.AbstractResource;
+import org.nuxeo.ecm.webengine.model.impl.ResourceTypeImpl;
+import org.nuxeo.runtime.api.Framework;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Endpoint to retrieve OAuth2 authentication data
+ * @since 8.4
+ */
+@WebObject(type = "oauth2")
+public class OAuth2Object extends AbstractResource<ResourceTypeImpl> {
+
+    /**
+     * Retrieves oauth2 data for a given provider.
+     */
+    @GET
+    @Path("provider/{providerId}")
+    public Response getProvider(@PathParam("providerId") String providerId,
+                                @Context HttpServletRequest request) throws IOException, RestOperationException {
+
+        NuxeoOAuth2ServiceProvider provider = getProvider(providerId);
+
+        Map<String,Object> result = new HashMap<>();
+        result.put("serviceName", provider.getServiceName());
+        result.put("enabled", provider.isEnabled());
+        result.put("clientId", provider.getClientId());
+        result.put("scopes", provider.getScopes());
+        result.put("authorizationURL", provider.getClientId() == null ? null : provider.getAuthorizationUrl(request));
+        result.put("authorizationServerURL", provider.getAuthorizationServerURL());
+        result.put("tokenServerURL", provider.getTokenServerURL());
+
+        String username = request.getUserPrincipal().getName();
+        NuxeoOAuth2Token token = getToken(provider, username);
+        boolean hasToken = (token != null);
+        String login = hasToken ? token.getServiceLogin() : null;
+        result.put("hasToken", hasToken);
+        result.put("userId", login);
+
+        return buildResponse(Status.OK, result);
+    }
+
+    /**
+     * Retrieves a valid access token for a given provider and the current user.
+     * If expired, the token will be refreshed.
+     */
+    @GET
+    @Path("provider/{providerId}/token")
+    public Response getToken(@PathParam("providerId") String providerId,
+                             @Context HttpServletRequest request) throws IOException, RestOperationException {
+
+        NuxeoOAuth2ServiceProvider provider = getProvider(providerId);
+
+        String username = request.getUserPrincipal().getName();
+        NuxeoOAuth2Token token = getToken(provider, username);
```

Comment:
Not necessarily, because the username we have so far is a nuxeo principal and the loadCredential method will expect a service user (email) if the provider extends AbstractOAuth2UserEmailProvider.